### DISCUSSION TITLE
Unable to upload files Mastodon 3.5.3

### DISCUSSION THREAD

**Discussion by hannahpi** [mastodon/mastodon#21582] [Q&A] [answered]
Hello I'm trying to set up a Mastodon 3.5.3 instance using DigitalOcean and use Amazon AWS S3.  I checked the journalctl and see no errors whatsoever (which was surprising to me).  I'm not able to confirm the link that it is trying to access to make sure my AWS settings are correct (at this point I've messed with them quite a bit without success).  I get the feeling that I don't have access or the S3 is not active or that there is some sort of access issue going on but I'm not quite sure how to confirm it.  I was going to try the s3cmd tool I came across in the posts but I don't know that I want to enable encryption (and if I do enable encryption how to get the key for it).  

I saw similar discussions but for one of them the config did not have the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` configured.  I'm also getting no errors from the journal and everything that might be related seems to be getting errors.  

```
S3_ENABLED=true
S3_PROTOCOL=https
S3_BUCKET=****-mastodon-bucket
S3_REGION=us-east-2
S3_HOSTNAME=s3-us-east-2.amazonaws.com
AWS_ACCESS_KEY_ID=****
AWS_SECRET_ACCESS_KEY=****
```
Incidentally I checked my user and it shows that it was last used an hour ago.  So that's a bit strange.  Would it show activity if I had the key wrong? 
<img width="130" alt="image" src="https://user-images.githubusercontent.com/1085232/203855773-8758bce1-6a5a-43fb-822b-780921b5ba15.png">

I've continued to research this and I've been able to confirm hits on Amazon's side with regards to access.  Gave it a new key.  It shows recent usage for the particular key.  I tried to configure a _bad_ key in Mastodon and noticed that it still didn't generate errors (at least in the journal).  The log folder in /home/mastodon/live is empty and nginx logs show activity and some errors.  I noticed there was a bad SSL handshake but the IP did not look relevant to my activity (or AWS).  Anyway I'm at a bit of a loss right now.  I was able to upload an object (via the S3 console) to the bucket--obviously that worked.

**Comment by NSFWHubSocial** [comment, marked as answer]
I'm still learning myself so please bear with me but you might want to checkout the following video. 9 minutes and 4 seconds into the video, a group is created in IAM and granted a AmazonS3FullAccess policy. Then a user is created with programmatic access:

https://youtu.be/dmQsaooyKyM?t=544

The video should start at the correct location. You might want to check IAM to see if you have created a group with a AmazonS3FullAccess policy and user with programmatic access.

**Reply by hannahpi** [reply]
Hello thanks for looking into this for me.  I was actually using this video as a guide.  I had my AWS account for quite a while so I wasn't sure if there was something that could have been set that would block the S3 from the root account.  Today I installed the AWS s3 command line and tried a few commands, I think I mentioned before that I uploaded a file so I wanted to see if I could see it from the aws s3 tool given the access key/secret.  I was able to confirm that I can see the object so I should at least have read access.  I did give the AmazonS3FullAccess to the user in question.

I referred to this article on re:post: https://repost.aws/questions/QUHpyJb2yKTiao-ZgOqYlxmQ/i-cant-use-my-s-3-buckets

```
>aws s3api head-object --bucket ****-mastodon-bucket --key masto.jpg
{
    "AcceptRanges": "bytes",
    "LastModified": "2022-11-24T18:57:13+00:00",
    "ContentLength": 7114,
    "ETag": "\"****\"",
    "ContentType": "image/jpeg",
    "Metadata": {}
}
```
I checked `journalctl -u mastodon-web` and I now see some errors.  Never thought I'd be excited to see some error messages.  Anyway this is the error I'm getting.
```
 method=POST path=/api/v2/media format=html controller=Api::V2::MediaController action=create status=500 error='Aws::S3::Errors::AccessControlListNotSupported: The bucket does not allow ACLs' duration=174.29 view=0.00 db=2.10

Aws::S3::Errors::AccessControlListNotSupported (The bucket does not allow ACLs):
```
I'm pretty sure I disabled ACLs so... I guess I need to re-enable them?  I just enabled Access Control Lists and now uploads are working.  Not sure if this was covered in that video so maybe I'll go post something under that video.  Thank you!
<img width="265" alt="image" src="https://user-images.githubusercontent.com/1085232/204057358-4ef45fca-d054-41ab-8fd2-a8d074a23b0e.png">

**Reply by hannahpi** [reply]
Also wanted to follow up and mention that I did have a mistake on my original config.  Most notably I had `s3-us` and this version has `s3.us` I added the alias host but not sure if that's required or not.  I would have expected it to be able to build this given the other information.
So the following is my current s3 config:
```
S3_ENABLED=true
S3_PROTOCOL=https
S3_BUCKET=****-mastodon-bucket
S3_REGION=us-east-2
S3_HOSTNAME=s3.us-east-2.amazonaws.com
S3_ALIAS_HOST=****-mastodon-bucket.s3.us-east-2.amazonaws.com
```
